### DISCUSSION TITLE
Agents vs Agents_By_Type

### DISCUSSION THREAD

**Discussion by rboone2020** [mesa/mesa#1883] [Q&A] [answered]
Folks,

Your help in improving my understanding would be appreciated.   If I use:

for agent in self.schedule.agents:

I get an iteration where "agent" is an **agent** in the scheduler as one would imagine.  But if I use:

for agent in self.schedule.agents_by_type[Sheep]:

I instead get "agent" as an **integer**.  The difference seems odd to me.   I've used the first case and then a filter to get to sheep, say, but is there a more direct route?  

Thanks

**Comment by rht** [comment, marked as answer]
`schedule.agents_by_type` is a dictionary of dictionaries. You will have to do this instead
```python
for agent_id, agent in self.schedule.agents_by_type[Sheep].items():
```

**Reply by rboone2020** [reply]
OK, thanks.  Never would have gotten that without your help.

**Comment by EwoutH** [comment]
For future reference, extensive work has been done on Agent management in Mesa. This should be a lot easier from the next release. See:

- https://github.com/projectmesa/mesa/pull/1894
- https://github.com/projectmesa/mesa/pull/1916
